proof/refine/ARM/Syscall_R.thy

```diff
@@ -2677,6 +2593,400 @@ proof -
+         apply (wpsimp wp: check_budget_restart_false check_budget_restart_true')
+        apply (wpsimp wp: checkBudgetRestart_false checkBudgetRestart_true)
+       apply (clarsimp dest!: active_from_running
+                        simp: cur_sc_offset_ready_def current_time_bounded_strengthen pred_map_def
+                              ct_in_state_weaken[OF _ active_activatable])
+      apply clarsimp
+     apply (wpsimp wp: update_time_stamp_current_time_bounded_5
+                       update_time_stamp_cur_sc_offset_ready_cs)
+    apply wpsimp
+   apply clarsimp+
+  done
+
+lemma handleEvent_corres:
+  "corres (dc \<oplus> dc)
+     (einvs and (\<lambda>s. event \<noteq> Interrupt \<longrightarrow> ct_running s)
+      and (\<lambda>s. scheduler_action s = resume_cur_thread)
+      and valid_machine_time and current_time_bounded 5
+      and consumed_time_bounded and cur_sc_active and (ct_active or ct_idle)
+      and ct_not_in_release_q and ct_not_queued
+      and (\<lambda>s. cur_sc_offset_ready (consumed_time s) s)
+      and (\<lambda>s. cur_sc_offset_sufficient (consumed_time s) s))
+     (invs' and (\<lambda>s. event \<noteq> Interrupt \<longrightarrow> ct_running' s) and
+      (\<lambda>s. vs_valid_duplicates' (ksPSpace s)) and
+      (\<lambda>s. ksSchedulerAction s = ResumeCurrentThread))
+     (handle_event event) (handleEvent event)"
+  (is "corres _ (?P and ?ready and _) ?P' _ _")
+proof -
+  have hw:
+    "\<And>isBlocking canGrant.
+     corres dc (einvs and ct_running and valid_machine_time and current_time_bounded 2
+                and ct_released and (\<lambda>s. scheduler_action s = resume_cur_thread)
+                and ct_not_in_release_q and ct_not_queued)
+               (invs' and ct_running'
+                      and (\<lambda>s. ksSchedulerAction s = ResumeCurrentThread))
+               (handle_recv isBlocking canGrant) (handleRecv isBlocking canGrant)"
+    apply add_cur_tcb'
+    apply add_ct_not_inQ
+    apply (rule corres_guard_imp [OF handleRecv_isBlocking_corres])
+     apply (fastforce simp: ct_in_state_def ct_in_state'_def
+                     elim!: st_tcb_weakenE pred_tcb'_weakenE
+                      dest: ct_not_ksQ)+
+    done
+  show ?thesis
+    apply (rule corres_cross_add_abs_guard[where Q="\<lambda>s. event \<noteq> Interrupt \<longrightarrow> ct_released s"])
+     apply (simp only: schact_is_rct_def[symmetric])
+     apply clarsimp
+     apply (frule (1) invs_strengthen_cur_sc_tcb_are_bound[OF _ invs_cur_sc_tcb])
+      apply (clarsimp simp: invs_def valid_state_def valid_pspace_def)
+     apply (erule (5) schact_is_rct_ct_released)
+     apply (erule (2) cur_sc_not_idle_sc_ptr')
+    apply (case_tac event)
+         apply (simp_all add: handleEvent_def)
+         apply (rename_tac syscall)
+         apply (rule updateTimeStamp_checkBudgetRestart_helperE)
+         apply (case_tac syscall; simp)
+                   apply (auto intro: corres_guard_imp[OF handleSend_corres]
```

Comment:
This `auto` needs to be removed. Maybe an Eisbach method, similar to the Eisbach methods we have in `DetSchedSchedule` for the `handle_event` Hoare triples, would be a good idea (though they are annoying to fix when they break...)

Reply:
This auto doesn’t need to be removed. This proof has pretty much (but not exactly) the same structure as the master one up to this point. This auto discharges all the subgoals that it is applied to. This auto may be removed if you like altogether with the master one, I suppose, but that would be total restructuring the proofs which is beyond the purpose of the current PR (which is to fix the sorries as the first priority).

Reply:
In terms of our style rules, this `auto` is permissible, yes. The important part is that it is only applied to goals that it solves, which seems to be the case here.